### DISCUSSION TITLE
Add select_next and select_previous services to Select

### DISCUSSION THREAD

**Discussion by avee87** [home-assistant/architecture#590] [Entity Models] [answered]
Currently the only service supported by [Select](https://www.home-assistant.io/integrations/select/) is `select_option`. This is nice but in some cases it may be useful to have an ability to loop through them consecutively. It is technically possible right now but requires a bunch of manual work and is generally not user-friendly.

One use case I have in mind is looping through WLED presets but most likely there are others.
Also note that [Input Select](https://www.home-assistant.io/integrations/input_select/) already has these and a bunch of others.

This could be done entirely in the Select code so it won't require writing any additional code in specific integrations using Select. 

Here's a PR implementing this functionality: https://github.com/home-assistant/core/pull/53485

**Comment by frenck** [comment]
So a couple of things with this, IMHO:

- Input select is a manual user list, that is not integrated. While closely related, they are very different in scope.
- Entity core components are minimal on purpose. Integrations can provide their own services if needed. But I can see the use case. However, I would argue that select entities are probably mainly going to represent a settings/option for a specific device. In those cases, these methods might be useless (or maybe even harmful?)
- The `input_select` provides these services:

   - `select_first`
   - `select_last`
   - `select_next`
   - `select_previous`

  Any specific reason why you are only requesting `select_next` and `select_previous`? 

The questions that we have to ask ourselves for this:

- Should these be added to the `select` entity?
- Can these options potentially be problematic for integrations?

**Reply by avee87** [reply]
> I would argue that select entities are probably mainly going to represent a settings/option for a specific device. In those cases, these methods might be useless (or maybe even harmful?)

The same argument can be made about the same methods on `input_select` in many cases. I don't see how they would be harmful unless user wants to harm themselves by using them incorrectly.

> Any specific reason why you are only requesting select_next and select_previous?

No particular reason, I'm happy to include them as well if someone needs them.

**Reply by frenck** [reply]
> The same argument can be made about the same methods on input_select in many cases.

Input select doesn't control anything or integrate with anything. It is just an state, the actions on it are NOOPs, it is up to the user to automate with it. Select entities are provided by integrations and changing those do have consequences.

**Reply by ivlis** [reply]
@frenck select is an abstraction that other integrations implement. WLED changes the preset if select changes. I see no difference if the select state is chosen by a mouse in the UI or programmatically with a next or previous call. 

How can I change my WLED to the next preset? This is a common use case for WLED.

> It is just an state, the actions on it are NOOPs, it is up to the user to automate with it.

Well, select implements `select.select_state`. I'm really not getting how `select_next` will be different from a ridiculous automation that reads all states and selects the next one. (Even if this is possible without custom components).

You deprecated `wled.preset` service in order get everything under the roof of select. I can do the same with `select.select_option` as with `wled.preset`. Now you are saying that the integrations should implement it.  So what's the point of `select` if I still need to write an HTTP call to WLED? Or write `wled.preset_next`?

**Reply by ivlis** [reply]
However, I see a point that for some integration options are a set (not a list) and the particular order does not make sense for them. Although ( https://developers.home-assistant.io/docs/core/entity/select/ ) specifies options as a list, not as a set, implying a particular order.

Yes, there is a case if there is a harmful situation of selecting Option B after selecting Option A. but:
1. It's not preventable right now in the UI or with select_option
2. It can be prevented by an integration, for example here: https://github.com/home-assistant/core/blob/18bf0762b551d2469ba63aaa0e2c2aa6f1dbc2db/homeassistant/components/esphome/select.py#L41

**Comment by frenck** [comment, marked as answer]
We have recently discussed this request in a meeting and decided to fully sync up select & input_select.

This means, all services & capabilities are synced (except for the service for setting options, which is limited to the input_select of course).

<https://github.com/home-assistant/core/pull/87255>
<https://github.com/home-assistant/home-assistant.io/pull/26092>

../Frenck
